pkg/controller/trainingruntime_controller.go

```diff
@@ -0,0 +1,149 @@
+var _ TrainJobWatcher = (*TrainingRuntimeReconciler)(nil)
+
+func NewTrainingRuntimeReconciler(cli client.Client, recorder record.EventRecorder) *TrainingRuntimeReconciler {
+	return &TrainingRuntimeReconciler{
+		log:                      ctrl.Log.WithName("trainingruntime-controller"),
+		client:                   cli,
+		recorder:                 recorder,
+		nonRuntimeObjectUpdateCh: make(chan event.TypedGenericEvent[iter.Seq[types.NamespacedName]], updateChBuffer),
+	}
+}
+
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainingruntimes,verbs=create;get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainingruntimes/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainingruntimes/finalizers,verbs=get;update;patch
+
+func (r *TrainingRuntimeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var runtime trainer.TrainingRuntime
+	if err := r.client.Get(ctx, req.NamespacedName, &runtime); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log := ctrl.LoggerFrom(ctx).WithValues("trainingRuntime", klog.KObj(&runtime))
+	ctrl.LoggerInto(ctx, log)
+	log.V(2).Info("Reconciling TrainingRuntime")
+
+	trainJobs := &trainer.TrainJobList{}
+	if err := r.client.List(ctx, trainJobs, client.InNamespace(runtime.Namespace), client.MatchingFields{
+		idxer.TrainJobRuntimeRefKey: runtime.Name,
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+	if !ctrlutil.ContainsFinalizer(&runtime, constants.ResourceInUseFinalizer) && len(trainJobs.Items) != 0 {
+		ctrlutil.AddFinalizer(&runtime, constants.ResourceInUseFinalizer)
+		return ctrl.Result{}, r.client.Update(ctx, &runtime)
+	} else if !runtime.DeletionTimestamp.IsZero() && len(trainJobs.Items) == 0 {
+		ctrlutil.RemoveFinalizer(&runtime, constants.ResourceInUseFinalizer)
+		return ctrl.Result{}, r.client.Update(ctx, &runtime)
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *TrainingRuntimeReconciler) NotifyTrainJobUpdate(oldJob, newJob *trainer.TrainJob) {
+	var runtimeNSName *types.NamespacedName
+	switch {
+	case oldJob != nil && newJob != nil:
+		// UPDATE Event.
+		return
+	case oldJob == nil:
+		// CREATE Event.
+		if trainjob.RuntimeRefIsTrainingRuntime(newJob.Spec.RuntimeRef) {
+			runtimeNSName = &types.NamespacedName{Namespace: newJob.Namespace, Name: newJob.Spec.RuntimeRef.Name}
+		}
+	default:
+		// DELETE Event.
+		if trainjob.RuntimeRefIsTrainingRuntime(oldJob.Spec.RuntimeRef) {
+			runtimeNSName = &types.NamespacedName{Namespace: oldJob.Namespace, Name: oldJob.Spec.RuntimeRef.Name}
+		}
+	}
+	if runtimeNSName != nil {
+		r.nonRuntimeObjectUpdateCh <- event.TypedGenericEvent[iter.Seq[types.NamespacedName]]{
+			Object: slices.Values([]types.NamespacedName{*runtimeNSName}),
+		}
+	}
+}
+
+type nonRuntimeObjectHandler struct{}
+
+var _ handler.TypedEventHandler[iter.Seq[types.NamespacedName], reconcile.Request] = (*nonRuntimeObjectHandler)(nil)
```

Comment:
do we need it ?

Reply:
This triggers TrainingRuntime reconciliations by TrainJob events.